### PR TITLE
Cloud module exceptions

### PR DESCRIPTION
**Pull request by bretttegart**
Added boto exception types and exception decorators to Cloud module.
Added generic exception handling to boto connection function in order to make troubleshooting easier.
Corrected issue with Supervisor unit test by sorting generated list of services prior to execution.

### CODE REVIEW

File: treadmill/cli/__init__.py [hunk 1]
```diff
@@ -945,3 +946,16 @@ def handle_not_authorized(err):
 ]
 
 ON_REST_EXCEPTIONS = handle_exceptions(REST_EXCEPTIONS)
+
+CLOUD_EXCEPTIONS = [
+    (botoexceptions.DataNotFoundError, 'Unable to load data'),
+    (botoexceptions.EndpointConnectionError,
+        'Could not connect to endpoint URL'),
+    (botoexceptions.NoCredentialsError, "Unable to locate credentials"),
```
Comment (andreikeis) [inline treadmill/cli/__init__.py]: nit - we use single quotes everywhere.

File: treadmill/infra/connection.py [hunk 2]
```diff
@@ -31,6 +31,9 @@ def __init__(self, resource=constants.EC2):
...
     def __new__(cls, resource=constants.EC2):
-        return boto3.client(
-            resource, region_name=cls.context.region_name
-        )
+        try:
+            return boto3.client(
+                resource, region_name=cls.context.region_name
+            )
+        except Exception as err:
+            print("AWS Client Exception: {}".format(str(err)))
```
Comment (andreikeis) [inline treadmill/infra/connection.py]: Not sure what the intent of this - logging? If logging, we do:

_LOGGER = logging.getLogger(__name__) in the beginning of the file and then:

_LOGGER.exception('aws client exception')

File: treadmill/supervisor.py [hunk 3]
```diff
@@ -241,7 +241,7 @@ def get_state(svcroot):
 
 def _service_wait(svcroot, up_opt, any_all_opt, timeout=0, subset=None):
     """Given services directory, wait for services to be in given state."""
-    services = glob.glob(os.path.join(svcroot, '*'))
+    services = sorted(glob.glob(os.path.join(svcroot, '*')))
```
Comment (andreikeis) [inline treadmill/supervisor.py]: why is this needed?

Comment (bretttegart) [reply treadmill/supervisor.py]: When running the test suite on my Linux system, the test treadmill/tests/supervisor_test.py fails on line 117. The test creates two folders (/a, /b), then checks to see if they are loaded into the _service_wait function in chronological order (/a, /b). During my test run, they are loaded in reverse chronological order (/b, /a). 

I checked the glob module and it states that results are returned in arbitrary order. Because the test assumes results will be ordered, I think that other functions that use the _service_wait function might make that assumption as well. I checked the supervisor_test commit history and I see that the test has been altered recently from testing against a single folder to include testing against multiple folders. 

The sorted() function was the simplest solution I thought of to solve this issue.

File: treadmill/infra/connection.py (hunk 2 above)
Comment (bretttegart) [reply treadmill/infra/connection.py]: This is a UI change intended to solve a setup issue I noticed when getting started with Treadmill. If the default AWS region is not defined in the user's environment, the connection is aborted and the user is faced with the enigmatic "Unable to load plugin: Cloud" error. 
This occurs before logging in cloud.py is initialized and before the new AWS exception handling is loaded. You showed me where to find a stack trace to debug the issue when I encountered this.

My intent with this commit was to sidestep that process and print the AWS connection error message for the user to diagnose. This has the benefit of delivering the error to the person who can correct it- we don't control the AWS configuration from within Treadmill. I'm happy to change the code to use the logger instead, but I believe we would then lose the UI benefits- the user would have to read the app logs to discover the issue.

File: treadmill/supervisor.py (hunk 3 above)
Comment (snyaggarwal) [reply treadmill/supervisor.py]: Ya this code is flaky. It fails our build as well. But in Morgan ms branch supervisor code is updated (refactored) and this goes away. So we are waiting for the new sync (which will happen in a day or two)

File: treadmill/infra/connection.py (hunk 2 above)
Comment (andreikeis) [reply treadmill/infra/connection.py]: why isn't the exception (i am sure it is not "Exception" rather some high level exception) part of CLI_CLOUD_EXCEPTIONS list? 

the unable to load error happens on unhandled exceptions.... i do not see reason why can't we catch this properly.

I assume this code is used in cli and in long running processes. In this case, _LOGGER.debug or _LOGGER.exception is required, we do not print anything from anywhere.

I think it is best not to handle this at all at this level. For cli - it will be propagated and caught by exception handler, for sproc (long running) it will be unhanded and service will terminate as intended.
